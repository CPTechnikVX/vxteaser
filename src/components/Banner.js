import React          from 'react';
import PropTypes      from 'prop-types';
import ElementFactory from '../utils/ElementFactory';

export default class Banner extends React.PureComponent {

	constructor(props) {
		super(props);

		// bind
		this.onBannerClick = this.onBannerClick.bind(this);
	}

	onBannerClick = (event) => {
		event.preventDefault();
		event.stopPropagation();

		if (typeof this.props.onBannerClick === 'function') {
			this.props.onBannerClick();
		}
	}

	render() {
		const config          = this.props.config;
		const color           = this.props.config.color ? this.props.config.color : 'white';
		const backgroundColor = this.props.config.backgroundColor ? this.props.config.backgroundColor : '#262728';
		const windowWidth     = this.props.windowWidth;

		let BannerElement;
		let BannerContent;
		if (this.props.config.template === 'fixedHeight') {
			const fixedHeights = this.props.config.fixedHeights;

			BannerElement = ElementFactory.getBanner({color, backgroundColor, fixedHeights, windowWidth});
		}

		const childrenWithProps = React.Children.map(this.props.children,
			(child) => {
				return React.cloneElement(child, {
					windowWidth: windowWidth,
				});
			}
		);

		return (
			<BannerElement onClick={this.onBannerClick} className="teaser-banner">
				{childrenWithProps}
			</BannerElement>
		);

	}
}

Banner.propTypes = {
	config:        PropTypes.object,
	children:      PropTypes.node,
	onBannerClick: PropTypes.func,
	windowWidth:   PropTypes.number,
};

Banner.defaultProps = {
	config:      {},
	windowWidth: window.innerWidth,
};
