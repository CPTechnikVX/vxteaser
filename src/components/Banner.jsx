import React from 'react';

class Banner extends React.PureComponent {
	render() {
		const styleObj = {
			width:           '1300px',
			height:          '260px',
			backgroundColor: 'lightgray',
			background:      'no-repeat url("https://ph.vxnextgen.x/assets/img/teaser/teaser-top10-bonus-bg.jpg") top left',
		};

		return (<div style={styleObj}>Banner</div>);
	}
}

export default Banner;