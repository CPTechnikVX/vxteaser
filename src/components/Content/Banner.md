```html
<banner theme="light" link="https://www.visit-x.net/de/magazin/aktuelle-ausgabe/">
    <skewpanelgroup skewWidth="54%">
      <panel width="45%" valign="bottom">
        <row>
          <col>
            <headline type="sub">Februar 2018</headline>
          </col>
        </row>
        <row>
          <col>
            <headline type="main">VISIT-X Magazin</headline>
          </col>
        </row>
        <row>
          <col>
            <headline>Exklusive Interviews mit den Camgirls, Live Einblicke in ein S/M Studio und die Chance das Traumauto von LexyRoxx zu gewinnen!</headline>
          </col>
        </row>
        <row>
          <col>
            <button type="primary">Jetzt lesen</button>
          </col>
        </row>
      </panel>
    </skewpanelgroup>
</banner>
```
```html
// banner has background position contain - for tablet (=> view2)
<banner modifier="-view2--bg-contain">
...
</banner>
```

```html
// banner with light theme
<banner modifier="-theme-light">
...
</banner>
```

```html
// banner with special theme
<banner modifier="-theme-special">
...
</banner>
```

```html
// banner with a transparent theme
<banner modifier="-theme-transparent">
...
</banner>
```
