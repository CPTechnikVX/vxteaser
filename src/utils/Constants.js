const classNames = {
	TeaserBanner:             'teaser-banner',
	TeaserBannerNoSkew:       'teaser-banner teaser--has-no-skew',
	TeaserBannerContent:      'teaser-banner__content',
	TeaserBannerContentInner: 'teaser-banner__content__inner',
	TeaserButton:             'teaser-button',
	TeaserSuite:              'teaser-suite',
	TeaserSuiteItem:          'teaser-suite__item',
	TeaserPoint:              'teaser-point',
	TeaserPointContainer:     'teaser-point-container',
};

export {
	classNames,
};