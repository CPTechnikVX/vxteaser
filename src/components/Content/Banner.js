import React         from 'react';
import PropTypes     from 'prop-types';
import classnames    from 'classnames';
import Constants     from '../../utils/Constants';
import BannerContent from './BannerContent';
import ClickEvent    from '../../utils/ClickEvent';

/**
 * Main Banner component
 */
export default class Banner extends React.PureComponent {
	constructor(props) {
		super(props);

		// bind
		this.onClickFn = this.onClickFn.bind(this);
	}

	onClickFn(event) {
		event.preventDefault();
		event.stopPropagation();

		this.props.onClickFn(new ClickEvent(event, this, {
			link: this.props.link,
		}));
	}

	render() {
		const {children, config, modifier, theme, windowWidth} = this.props;

		if (config) {
			const classList = [];
			let videoUrl    = config.fixedHeights[0]['backgroundUrl'].endsWith('.mp4') ? config.fixedHeights[0]['backgroundUrl'] : '';
			const styleObj  = {
				backgroundImage: videoUrl ? "" : `url('${config.fixedHeights[0]['backgroundUrl']}')`, 
				height:          `${config.fixedHeights[1]['height']}px`,
			};
			
			if (windowWidth < config.fixedHeights[1]['greaterThan']) {
				videoUrl = config.fixedHeights[1]['backgroundUrl'].endsWith('.mp4') ? config.fixedHeights[1]['backgroundUrl'] : '';
				styleObj.backgroundImage = videoUrl ? "" : `url('${config.fixedHeights[1]['backgroundUrl']}')`;
				styleObj.height          = `${config.fixedHeights[0]['height']}px`;

				classList.push(Constants.ClassName.BannerNoSkew);
				classList.push('-view2');
			} else {
				classList.push(Constants.ClassName.Banner);
				classList.push('-view1');
			}

			if (theme) {
				classList.push('-theme-' + theme);
			}

			if (modifier) {
				classList.push(modifier);
			}

			return (
				<div className={classnames(classList)} style={styleObj} onClick={this.onClickFn} data-id={config.id}>
					{videoUrl && <video playsInline autoPlay muted loop style={{position: 'absolute', width: '100%', height: '100%', objectFit: 'cover'}}><source src={videoUrl} /></video>}
					<BannerContent config={config} windowWidth={windowWidth}>{children}</BannerContent>
				</div>
			);
		} else {
			return null;
		}
	}
}

Banner.propTypes = {
	/** @ignore */
	config:      PropTypes.object,
	/** @ignore */
	children:    PropTypes.node,
	modifier:    PropTypes.string,
	/** @ignore */
	onClickFn:   PropTypes.func,
	link:        PropTypes.string,
	theme:       PropTypes.string,
	/** @ignore */
	windowWidth: PropTypes.number,
};
