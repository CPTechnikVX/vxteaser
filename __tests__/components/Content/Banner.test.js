import React            from 'react';
import {render, screen} from '@testing-library/react';
import Banner           from '../../../src/components/Content/Banner';
import Constants        from '../../../src/utils/Constants';

describe('render banner', () => {
    const config = {
        fixedHeights: [
            {
                backgroundUrl: 'url0',
                height:        500,
            },
            {
                backgroundUrl: 'url1',
                height:        260,
                greaterThan:   1200,
            },
        ],
    };

    test('empty banner', () => {
        const {container} = render(
                <Banner />
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Banner renders correctly', () => {
        const config = {
            fixedHeights: [
                {backgroundUrl: 'test-url.jpg', height: 100},
                {height: 200}
            ]
        };

        render(<Banner config={config} />);

        // Check if the image is rendered with correct src and style
        const image = screen.getByRole('img');
        expect(image).toHaveAttribute('src', config.fixedHeights[0].backgroundUrl);
        expect(image).toHaveStyle({height: '100%'});

        // Check if the banner has the correct height
        const banner = screen.getByTestId('banner');
        expect(banner).toHaveStyle({height: '200px'});

        // Check if BannerContent is rendered
        expect(screen.getByTestId('banner-content')).toBeInTheDocument();

        // Check if the banner has the correct class
        expect(banner).toHaveClass(Constants.ClassName.Banner[0]);

        // Snapshot testing
        expect(banner).toMatchSnapshot();
    });

    test('banner content tablet', () => {
        render(<Banner config={config} windowWidth={768} />);

        const image = screen.getByRole('img');
        expect(image).toHaveAttribute('src', config.fixedHeights[1].backgroundUrl);

        const banner = screen.getByTestId('banner');
        expect(banner).toHaveClass(Constants.ClassName.BannerNoSkew);

        expect(screen.getByTestId('banner-content')).toBeInTheDocument();

        expect(banner).toMatchSnapshot();
    });

    test('banner theme', () => {
        const {container} = render(
                <Banner config={config} theme={'white'} modifier="-view2--bg-auto" />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

});
