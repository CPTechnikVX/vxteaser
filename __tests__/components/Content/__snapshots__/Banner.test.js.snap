// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render banner Banner renders correctly 1`] = `
<div
  class="vxteaser-banner vxteaser--has-skew -view1"
  data-testid="banner"
  style="height: 200px;"
>
  <img
    loading="lazy"
    src="test-url.jpg"
    style="width: 100%; height: 100%; object-fit: cover; position: absolute; top: 0px;"
  />
  <div
    class="vxteaser-banner__content"
    data-testid="banner-content"
  >
    <div
      class="vxteaser-banner__content__inner"
    />
  </div>
</div>
`;

exports[`render banner banner content tablet 1`] = `
<div
  class="vxteaser-banner vxteaser--has-no-skew -view2"
  data-testid="banner"
  style="height: 500px;"
>
  <img
    loading="lazy"
    src="url1"
    style="width: 100%; height: 260px; object-fit: cover; position: absolute; top: 0px;"
  />
  <div
    class="vxteaser-banner__content"
    data-testid="banner-content"
    style="height: 250px; margin-top: 250px; text-align: center;"
  >
    <div
      class="vxteaser-banner__content__inner"
    />
  </div>
</div>
`;

exports[`render banner banner theme 1`] = `
<div
  class="vxteaser-banner vxteaser--has-skew -view1 -theme-white -view2--bg-auto"
  data-testid="banner"
  style="height: 260px;"
>
  <img
    loading="lazy"
    src="url0"
    style="width: 100%; height: 100%; object-fit: cover; position: absolute; top: 0px;"
  />
  <div
    class="vxteaser-banner__content"
    data-testid="banner-content"
  >
    <div
      class="vxteaser-banner__content__inner"
    />
  </div>
</div>
`;

exports[`render banner empty banner 1`] = `null`;
